release.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Release new version of artman.
+
+This script updates SHA hashes of latest commits for both gapic-generator
+and googleapis, and updates version number for subsequent (auto-)release.
+
+Usage:
+$ python3 release.py googleapis [sha]      # update googleapis commit hash
+$ python3 release.py gapic-generator [sha] # update gapic-generator commit hash
+$ python3 release.py 1.12.4                # update artman version
+"""
+
+import os.path
+import re
+import requests
+import sys
+from ruamel.yaml import YAML
+
+dockerfile_googleapis_regex = r'^(ENV GOOGLEAPIS_HASH )\S+()'
+dockerfile_gapic_generator_regex = r'^(ENV GAPIC_GENERATOR_HASH )\S+()'
+dockerfile_version_regex = r'^(ENV ARTMAN_VERSION )\S+()'
+setup_py_version_regex = r"^(current_version = ')\S+(')"
+
+github_gapic_generator_project = 'googleapis/gapic-generator'
+github_googleapis_project = 'googleapis/googleapis'
+
+
+def usage():
+    print('Usage:')
+    print(
+        '$ python3 release.py googleapis [sha]      # update googleapis commit hash'
+    )
+    print(
+        '$ python3 release.py gapic-generator [sha] # update gapic-generator commit hash'
+    )
+    print('$ python3 release.py 1.12.4                # update artman version')
+
+
+def get_latest_commit_sha(project):
+    url = 'https://api.github.com/repos/%s/branches/master' % project
+    # try to get GitHub credentials from config
+    auth = None
+    artman_config = os.path.join(os.environ['HOME'], '.artman', 'config.yaml')
+    try:
+        yaml = YAML()
+        with open(artman_config) as config_yaml:
+            config = yaml.load(config_yaml)
+            github_username = config['github']['username']
+            github_token = config['github']['token']
+            auth = requests.auth.HTTPBasicAuth(github_username, github_token)
+    except:
+        print('Cannot get GitHub token from %s, trying anonymously...' %
+              artman_config)
```

Comment:
Why can't we always do it anonymously?

Reply:
Because of the rate limit. When running on gLinux machine, it's very likely you'll get 403 from GitHub API because someone has eaten 60 queries per hour per IP that are allowed for anonymous queries.

```sh
$ curl 2>/dev/null https://api.github.com/rate_limit | jq .resources.core
{
  "limit": 60,
  "remaining": 31,
  "reset": 1531350123
}
```

Authenticated users get higher limits. It's not that important so the script is just doing its best trying to locate a config in the default place and does not fail if there is no config, or no token there.